[qgis]Add new port

longhuan2018 opened the pull request:
new port for #7761

longhuan2018 committed:
new port qgis

longhuan2018 committed:
check empty directories

Check for empty directories before file REMOVE_RECURSE

longhuan2018 committed:
support server feature

ras0219-msft commented on ports/qgis/portfile.cmake:
```diff
@@ -0,0 +1,387 @@
+
+# Handle copyright
+file(COPY ${SOURCE_PATH}/COPYING DESTINATION ${CURRENT_PACKAGES_DIR}/share/qgis)
+file(RENAME ${CURRENT_PACKAGES_DIR}/share/qgis/COPYING ${CURRENT_PACKAGES_DIR}/share/qgis/copyright)
```
https://github.com/Microsoft/vcpkg/blob/master/scripts/cmake/vcpkg_copy_tool_dependencies.cmake

longhuan2018 replied:
Ok, I will add vcpkg_copy_tool_dependencies

longhuan2018 replied:
I extended vcpkg_copy_tool_dependencies because I found that it did not fully meet my needs

longhuan2018 commented on ports/qgis/portfile.cmake:
```diff
@@ -0,0 +1,387 @@
+
+##############################################################################
+#Install pip
+if(NOT EXISTS "${PYTHON_PATH}/Scripts/pip.exe")
```
I will run under Ubuntu and modify it to work properly.

longhuan2018 replied:
I tried to compile the code under ubuntu and made it complete the config step, but it failed at build time and I haven't found a solution

NancyLi1013 commented on ports/qgis/CONTROL:
```diff
@@ -0,0 +1,17 @@
+Source: qgis
+Version: final-3_8_2
+Description:  QGIS is a free, open source, cross platform (lin/win/mac) geographical information system (GIS) 
+Build-Depends: expat, hdf5, libxml2, libzip, zlib, libspatialindex, libspatialite, gdal, opencl, exiv2, gsl, qt5, qwt, qca, qscintilla, qtkeychain, python3
```
Could you please add homepage here?

longhuan2018 replied:
I have completed the modification

NancyLi1013 commented on ports/qgis/portfile.cmake:
```diff
@@ -0,0 +1,387 @@
+include(vcpkg_common_functions)
```
`include(vcpkg_common_functions)` is no longer needed.
Could you please remove this line?

longhuan2018 replied:
I have completed the modification

NancyLi1013 commented on ports/qgis/portfile.cmake:
```diff
@@ -0,0 +1,387 @@
+    REF final-3_8_2
+    SHA512   09c54cc9afc21f792cedd413f0f752480ecabae37c54625c55f45473b1241893b4e55ede38894794d08819c3107cb07e2879ccce909a74e1295d17e049d7c737
+    HEAD_REF master
+	PATCHES
```
Could you please align-left `PATCHES `with above keywords(`HEAD_REF ` `SHA512 `...)?

longhuan2018 replied:
I have completed the modification

NancyLi1013 commented on ports/qgis/portfile.cmake:
```diff
@@ -0,0 +1,387 @@
+get_filename_component(PYTHON_PATH ${PYTHON3} PATH)
+set(ENV{PATH} "$ENV{PATH};${PYTHON_PATH};${PYTHON_PATH}/Scripts")
+set(PYTHONHOME ${PYTHON_PATH})
+
```
We can updated these as the following codes:
```
vcpkg_find_acquire_program(PYTHON3)
get_filename_component(PYTHON3_PATH ${PYTHON3} DIRECTORY)
vcpkg_add_to_path(${PYTHON3_PATH})
```

longhuan2018 replied:
I have completed the modification

NancyLi1013 commented on ports/qgis/portfile.cmake:
```diff
@@ -0,0 +1,387 @@
+
+# Handle copyright
+file(COPY ${SOURCE_PATH}/COPYING DESTINATION ${CURRENT_PACKAGES_DIR}/share/qgis)
+file(RENAME ${CURRENT_PACKAGES_DIR}/share/qgis/COPYING ${CURRENT_PACKAGES_DIR}/share/qgis/copyright)
```
Could you update 
```
file(COPY ${SOURCE_PATH}/COPYING DESTINATION ${CURRENT_PACKAGES_DIR}/share/qgis)
file(RENAME ${CURRENT_PACKAGES_DIR}/share/qgis/COPYING ${CURRENT_PACKAGES_DIR}/share/qgis/copyright)
```
as
`file(INSTALL ${SOURCE_PATH}/COPYING DESTINATION ${CURRENT_PACKAGES_DIR}/share/${PORT} RENAME copyright)`?

longhuan2018 replied:
I have completed the modification

NancyLi1013 commented on ports/qgis/portfile.cmake:
```diff
@@ -0,0 +1,387 @@
+		# In vcpkg, qca's library name is qca, but qgis defaults to qca-qt5 or qca2-qt5, so add qca for easy searching
+		qca.patch
+		#The postgres, ogr, wms, arcgisrest providers plugin and oauth2 will be compiled into a static library and a dynamic library at the same time. It is possible to use the moc to generate the same file at compile time, which will cause compilation errors. So remove the static library build.
+		static.patch
```
Could you please submit this patch to upstream?

longhuan2018 replied:
Compiling with the latest code, the static library problem disappeared, so I removed the static patch package

longhuan2018 commented:
Sorry, I will modify it after finishing the leave

longhuan2018 committed:
Merge pull request #5 from microsoft/master

update from vcpkg

vicroms commented:
/azp run

NancyLi1013 commented:
@longhuan2018 thanks for your update.
I noticed that this PR failed on `x86-windows` and `x64-windows` triplets.
The error info like this:
```
9>C:\vsts\_work\4\s\buildtrees\qgis\src\inal-3_8_2-055fb45c99\src\core\geometry\qgsinternalgeometryengine.cpp(924,48): warning C4267:  'argument': conversion from 'size_t' to 'int', possible loss of data [C:\vsts\_work\4\s\buildtrees\qgis\x64-windows-dbg\src\core\qgis_core.vcxproj]
C:\vsts\_work\4\s\buildtrees\qgis\src\inal-3_8_2-055fb45c99\src\core\geometry\qgsinternalgeometryengine.cpp(924,48): warning C4267: cl2.hpp: CL_HPP_TARGET_OPENCL_VERSION is not a valid value (100, 110, 120, 200 or 210). It will be set to 210 [C:\vsts\_work\4\s\buildtrees\qgis\x64-windows-dbg\src\core\qgis_core.vcxproj]
C:\vsts\_work\4\s\buildtrees\qgis\src\inal-3_8_2-055fb45c99\src\core\geometry\qgsinternalgeometryengine.cpp(924,48): warning C4267: CL_TARGET_OPENCL_VERSION is already defined as is lower than CL_HPP_TARGET_OPENCL_VERSION [C:\vsts\_work\4\s\buildtrees\qgis\x64-windows-dbg\src\core\qgis_core.vcxproj]
     9>C:\vsts\_work\4\s\installed\x64-windows\include\CL/cl2.hpp(1392,1): error C2065:  'CL_PLATFORM_HOST_TIMER_RESOLUTION': undeclared identifier (compiling source file C:\vsts\_work\4\s\buildtrees\qgis\src\inal-3_8_2-055fb45c99\src\core\qgsopenclutils.cpp) [C:\vsts\_work\4\s\buildtrees\qgis\x64-windows-dbg\src\core\qgis_core.vcxproj]
     9>C:\vsts\_work\4\s\installed\x64-windows\include\CL/cl2.hpp(1392,1): error C2975:  'Name': invalid template argument for 'cl::detail::param_traits', expected compile-time constant expression (compiling source file C:\vsts\_work\4\s\buildtrees\qgis\src\inal-3_8_2-055fb45c99\src\core\qgsopenclutils.cpp) [C:\vsts\_work\4\s\buildtrees\qgis\x64-windows-dbg\src\core\qgis_core.vcxproj]
       C:\vsts\_work\4\s\installed\x64-windows\include\CL/cl2.hpp(1369): message :  see declaration of 'Name' (compiling source file C:\vsts\_work\4\s\buildtrees\qgis\src\inal-3_8_2-055fb45c99\src\core\qgsopenclutils.cpp) [C:\vsts\_work\4\s\buildtrees\qgis\x64-windows-dbg\src\core\qgis_core.vcxproj]
     9>C:\vsts\_work\4\s\installed\x64-windows\include\CL/cl2.hpp(1392,1): error C2065:  'CL_PROGRAM_IL': undeclared identifier (compiling source file C:\vsts\_work\4\s\buildtrees\qgis\src\inal-3_8_2-055fb45c99\src\core\qgsopenclutils.cpp) [C:\vsts\_work\4\s\buildtrees\qgis\x64-windows-dbg\src\core\qgis_core.vcxproj]
     9>C:\vsts\_work\4\s\installed\x64-windows\include\CL/cl2.hpp(1392,1): error C2975:  'Name': invalid template argument for 'cl::detail::param_traits', expected compile-time constant expression (compiling source file C:\vsts\_work\4\s\buildtrees\qgis\src\inal-3_8_2-055fb45c99\src\core\qgsopenclutils.cpp) [C:\vsts\_work\4\s\buildtrees\qgis\x64-windows-dbg\src\core\qgis_core.vcxproj]
       C:\vsts\_work\4\s\installed\x64-windows\include\CL/cl2.hpp(1369): message :  see declaration of 'Name' (compiling source file C:\vsts\_work\4\s\buildtrees\qgis\src\inal-3_8_2-055fb45c99\src\core\qgsopenclutils.cpp) [C:\vsts\_work\4\s\buildtrees\qgis\x64-windows-dbg\src\core\qgis_core.vcxproj]
```
Could you take a look and try to fix the regressions?

NancyLi1013 commented:
Hi @longhuan2018 
Could you please address the review suggestions and regressions?

longhuan2018 commented:
Hi @NancyLi1013
Sorry, because I'm in Wuhan, all the logistics speed is a bit slow now. When the new computer arrives, I will start processing.

NancyLi1013 commented:
Ok, thanks for your update @longhuan2018.

longhuan2018 committed:
Merge pull request #7 from microsoft/master

merge from vcpkg

longhuan2018 committed:
[qgis qscintilla] Update to 3.12.0

Because qgis requires a higher version of qscintilla, so update qscintilla to 2.11.4

longhuan2018 commented:
> @longhuan2018 thanks for your update.
> I noticed that this PR failed on `x86-windows` and `x64-windows` triplets.
> The error info like this:
> 
> ```
> 9>C:\vsts\_work\4\s\buildtrees\qgis\src\inal-3_8_2-055fb45c99\src\core\geometry\qgsinternalgeometryengine.cpp(924,48): warning C4267:  'argument': conversion from 'size_t' to 'int', possible loss of data [C:\vsts\_work\4\s\buildtrees\qgis\x64-windows-dbg\src\core\qgis_core.vcxproj]
> C:\vsts\_work\4\s\buildtrees\qgis\src\inal-3_8_2-055fb45c99\src\core\geometry\qgsinternalgeometryengine.cpp(924,48): warning C4267: cl2.hpp: CL_HPP_TARGET_OPENCL_VERSION is not a valid value (100, 110, 120, 200 or 210). It will be set to 210 [C:\vsts\_work\4\s\buildtrees\qgis\x64-windows-dbg\src\core\qgis_core.vcxproj]
> C:\vsts\_work\4\s\buildtrees\qgis\src\inal-3_8_2-055fb45c99\src\core\geometry\qgsinternalgeometryengine.cpp(924,48): warning C4267: CL_TARGET_OPENCL_VERSION is already defined as is lower than CL_HPP_TARGET_OPENCL_VERSION [C:\vsts\_work\4\s\buildtrees\qgis\x64-windows-dbg\src\core\qgis_core.vcxproj]
>      9>C:\vsts\_work\4\s\installed\x64-windows\include\CL/cl2.hpp(1392,1): error C2065:  'CL_PLATFORM_HOST_TIMER_RESOLUTION': undeclared identifier (compiling source file C:\vsts\_work\4\s\buildtrees\qgis\src\inal-3_8_2-055fb45c99\src\core\qgsopenclutils.cpp) [C:\vsts\_work\4\s\buildtrees\qgis\x64-windows-dbg\src\core\qgis_core.vcxproj]
>      9>C:\vsts\_work\4\s\installed\x64-windows\include\CL/cl2.hpp(1392,1): error C2975:  'Name': invalid template argument for 'cl::detail::param_traits', expected compile-time constant expression (compiling source file C:\vsts\_work\4\s\buildtrees\qgis\src\inal-3_8_2-055fb45c99\src\core\qgsopenclutils.cpp) [C:\vsts\_work\4\s\buildtrees\qgis\x64-windows-dbg\src\core\qgis_core.vcxproj]
>        C:\vsts\_work\4\s\installed\x64-windows\include\CL/cl2.hpp(1369): message :  see declaration of 'Name' (compiling source file C:\vsts\_work\4\s\buildtrees\qgis\src\inal-3_8_2-055fb45c99\src\core\qgsopenclutils.cpp) [C:\vsts\_work\4\s\buildtrees\qgis\x64-windows-dbg\src\core\qgis_core.vcxproj]
>      9>C:\vsts\_work\4\s\installed\x64-windows\include\CL/cl2.hpp(1392,1): error C2065:  'CL_PROGRAM_IL': undeclared identifier (compiling source file C:\vsts\_work\4\s\buildtrees\qgis\src\inal-3_8_2-055fb45c99\src\core\qgsopenclutils.cpp) [C:\vsts\_work\4\s\buildtrees\qgis\x64-windows-dbg\src\core\qgis_core.vcxproj]
>      9>C:\vsts\_work\4\s\installed\x64-windows\include\CL/cl2.hpp(1392,1): error C2975:  'Name': invalid template argument for 'cl::detail::param_traits', expected compile-time constant expression (compiling source file C:\vsts\_work\4\s\buildtrees\qgis\src\inal-3_8_2-055fb45c99\src\core\qgsopenclutils.cpp) [C:\vsts\_work\4\s\buildtrees\qgis\x64-windows-dbg\src\core\qgis_core.vcxproj]
>        C:\vsts\_work\4\s\installed\x64-windows\include\CL/cl2.hpp(1369): message :  see declaration of 'Name' (compiling source file C:\vsts\_work\4\s\buildtrees\qgis\src\inal-3_8_2-055fb45c99\src\core\qgsopenclutils.cpp) [C:\vsts\_work\4\s\buildtrees\qgis\x64-windows-dbg\src\core\qgis_core.vcxproj]
> ```
> 
> Could you take a look and try to fix the regressions?

I upgraded qgis to the latest version 3.12.0, and the problem is not reproduced after compiling with the latest vcpkg code, please try the latest uploaded code

longhuan2018 committed:
Fix file case

NancyLi1013 commented:
Thanks for your efforts.
It seems that there are still some regresssions. Could you look into them?

longhuan2018 committed:
add qt-latest feature

longhuan2018 commented:
> Thanks for your efforts.
> It seems that there are still some regresssions. Could you look into them?

It may be a problem with the version of qt. When I tested it on my computer, I used qt [latest]. I tried to add different versions of qt support to see. If it hasn't been resolved, I will try to solve it tomorrow

longhuan2018 commented:
qgis requires a higher version of qscintilla, so I upgraded qscintilla at the same time, but the build prompt for qscintilla failed under linux
```
2020-03-09T13:50:39.3621908Z ##[section]Starting: Fail pipeline on error
2020-03-09T13:50:39.3624905Z ==============================================================================
2020-03-09T13:50:39.3625226Z Task         : Bash
2020-03-09T13:50:39.3625385Z Description  : Run a Bash script on macOS, Linux, or Windows
2020-03-09T13:50:39.3625572Z Version      : 3.163.1
2020-03-09T13:50:39.3625713Z Author       : Microsoft Corporation
2020-03-09T13:50:39.3625849Z Help         : https://docs.microsoft.com/azure/devops/pipelines/tasks/utility/bash
2020-03-09T13:50:39.3626168Z ==============================================================================
2020-03-09T13:50:39.4904736Z Generating script.
2020-03-09T13:50:39.4934419Z ========================== Starting Command Output ===========================
2020-03-09T13:50:39.4955261Z [command]/bin/bash --noprofile --norc /ci/myagent/_work/_temp/9d151576-2e6d-46be-8443-54cfe04c43f4.sh
2020-03-09T13:50:39.5054497Z 
2020-03-09T13:50:39.5056754Z See port build failure logs' artifact under summary tab for more information on failures
2020-03-09T13:50:39.5086324Z ##[error]Bash wrote one or more lines to the standard error stream.
2020-03-09T13:50:39.5087377Z ##[error]See port build failure logs' artifact under summary tab for more information on failures

2020-03-09T13:50:39.5114265Z ##[section]Finishing: Fail pipeline on error
```
```
There are no build logs for qscintilla build.
This is usually because the build failed early and outside of a task that is logged.
See the console output logs from vcpkg for more information on the failure.
If the console output of the qscintilla is missing you can trigger a rebuild
in the test system by making any whitespace change to one of the files under
the ports/qscintilla directory or by asking a member of the vcpkg team to remove the
tombstone for abi tag 3eb82c8eaea336c042361d1d0fe424ebedd25d2c
```

NancyLi1013 commented:
It seems that `qscintilla` has passed on x64-linux triplet.
Could you please try to remove `qscintilla:x64-linux` from` ci.baseline.txt`?

`##[error] PASSING, REMOVE FROM FAIL LIST: qscintilla:x64-linux (/ci/myagent/_work/1/s/scripts/ci.baseline.txt)`

NancyLi1013 commented:
/azp run

NancyLi1013 commented:
@longhuan2018 
Could you please resolve the conflicts?

longhuan2018 commented:
> @longhuan2018
> Could you please resolve the conflicts?

Ok， i will fix these problems

NancyLi1013 commented:
@longhuan2018 
Is work still being done for this PR?

NancyLi1013 commented:
@longhuan2018 
Please resolve the conflicts.

NancyLi1013 commented:
Thanks for the PR; we're closing this for now since there's been no response. If you'd like to continue working on it, please reopen and ping us!

longhuan2018 committed:
Revert qscintilla changes

longhuan2018 committed:
Merge pull request #15 from microsoft/master

merge from vcpkg

longhuan2018 committed:
update qgis version to 3.14.16

longhuan2018 commented:
I have re-updated the PR now. How do I reopen this PR?

NancyLi1013 commented:
I reopened this PR just now and you can continue working done for this PR now.

NancyLi1013 commented:
Please merge master branch to this PR before you work on since there are many changes.

longhuan2018 committed:
Merge pull request #17 from microsoft/master

merge from vcpkg

longhuan2018 committed:
add ltr debug-tools feature

longhuan2018 committed:
fix elseif bug

longhuan2018 committed:
enable qca build in linux and osx

longhuan2018 committed:
Test x86 compilation failure reason

longhuan2018 committed:
Merge pull request #22 from microsoft/master

merge from vcpkg

NancyLi1013 commented:
/azp run

NancyLi1013 commented:
The failures on x86-windows:
```
 QQmlComponent: Component is not ready
         file:///D:/buildtrees/qgis/x86-windows-dbg/src/quickgui/plugin/typelist.qml:3:1: plugin cannot be loaded for module "QgsQuick": Cannot load library D:\buildtrees\qgis\x86-windows-dbg\src\quickgui\plugin\QgsQuick\qgis_quick_plugin.dll: The specified procedure could not be found.
   127>C:\Program Files (x86)\Microsoft Visual Studio\2019\Enterprise\MSBuild\Microsoft\VC\v160\Microsoft.CppCommon.targets(149,5): error MSB3073: The command "setlocal [D:\buildtrees\qgis\x86-windows-dbg\src\quickgui\plugin\qgis_quick_plugin.vcxproj]
C:\Program Files (x86)\Microsoft Visual Studio\2019\Enterprise\MSBuild\Microsoft\VC\v160\Microsoft.CppCommon.targets(149,5): error MSB3073: cd D:\buildtrees\qgis\x86-windows-dbg\src\quickgui\plugin 
```
The failures on x64-linux:

`ninja: error: '/mnt/vcpkg-ci/installed/x64-linux/debug/lib/qt5keychaind.a', needed by 'output/lib/libqgis_core.so.3.14.16', missing and no known rule to make it`

@longhuan2018 Could you please take a look and try to fix them?

longhuan2018 commented:
> The failures on x86-windows:
> 
> ```
>  QQmlComponent: Component is not ready
>          file:///D:/buildtrees/qgis/x86-windows-dbg/src/quickgui/plugin/typelist.qml:3:1: plugin cannot be loaded for module "QgsQuick": Cannot load library D:\buildtrees\qgis\x86-windows-dbg\src\quickgui\plugin\QgsQuick\qgis_quick_plugin.dll: The specified procedure could not be found.
>    127>C:\Program Files (x86)\Microsoft Visual Studio\2019\Enterprise\MSBuild\Microsoft\VC\v160\Microsoft.CppCommon.targets(149,5): error MSB3073: The command "setlocal [D:\buildtrees\qgis\x86-windows-dbg\src\quickgui\plugin\qgis_quick_plugin.vcxproj]
> C:\Program Files (x86)\Microsoft Visual Studio\2019\Enterprise\MSBuild\Microsoft\VC\v160\Microsoft.CppCommon.targets(149,5): error MSB3073: cd D:\buildtrees\qgis\x86-windows-dbg\src\quickgui\plugin 
> ```
> 
> The failures on x64-linux:
> 
> `ninja: error: '/mnt/vcpkg-ci/installed/x64-linux/debug/lib/qt5keychaind.a', needed by 'output/lib/libqgis_core.so.3.14.16', missing and no known rule to make it`
> 
> @longhuan2018 Could you please take a look and try to fix them?

I am setting up a local environment for testing

strega-nil commented:
@longhuan2018 any news?

longhuan2018 committed:
Merge pull request #23 from microsoft/master

update from vcpkg

NancyLi1013 commented:
Failed on x64-windows and x86-windows:

```
CMake Error at ports/qgis/portfile.cmake:167 (file):
  file COPY cannot find "D:/installed/x86-windows/lib/python38.lib": No such
  file or directory.
Call Stack (most recent call first):
  scripts/ports.cmake:136 (include)
```
Failed on x64-linux:

`ninja: error: '/mnt/vcpkg-ci/installed/x64-linux/debug/lib/qt5keychaind.a', needed by 'output/lib/libqgis_core.so.3.14.16', missing and no known rule to make it
`
@longhuan2018 

Could you please look into these?

BillyONeal commented:
Closing due to inactivity. Please feel free to open a new PR if you want to continue.
